src/main/java/de/triology/cas/services/EtcdServicesManager.java

```diff
@@ -56,50 +51,66 @@
 
     private String fqdn;
 
+    private String stage;
+
     private final List<String> allowedAttributes;
 
-    private final JSONParser parser = new JSONParser();
+    private EtcdClient etcd;
 
-    private final ReadWriteLock lock = new ReentrantReadWriteLock();
+    public EtcdServicesManager(List<String> allowedAttributes, String stage) {
+        this.allowedAttributes = allowedAttributes;
+        this.stage = stage;
+    }
 
-    private final EtcdClient etcd;
+    private ConcurrentHashMap<Long, RegisteredService> getRegisteredServices() {
+        if (this.registeredServices.isEmpty()) {
+            initRegisteredServices();
+        }
+        return this.registeredServices;
+    }
 
-    public EtcdServicesManager(List<String> allowedAttributes, String stage) throws IOException, EtcdException, EtcdAuthenticationException, TimeoutException, ParseException {
-        this.allowedAttributes = allowedAttributes;
+    private void initRegisteredServices() {
         // depends on stage configured in cas.properties
-        if (!stage.equals("development")) {
+        if (!"development".equals(stage)) {
+            initProductionMode();
+
+        } else {
+            initDevelopmentMode();
+        }
+    }
+
+    private void initProductionMode() {
+        try {
             // check which stage is set in etcd
-            etcd = new EtcdClient(URI.create(getEtcdUri()));
-            try {
-                EtcdKeysResponse stageResponse = etcd.get("/config/_global/stage").send().get();
-                stage = stageResponse.getNode().getValue();
-            } catch (EtcdException ex) {
-                logger.warn("/config/_global/stage could not be read", ex);
-                stage = "production";
-            }
-            if (stage.equals("development")) {
-                initDevMode();
+            etcd = new EtcdClient(URI.create(EtcdRegistryUtils.getEtcdUri()));
+            EtcdKeysResponse stageResponse = etcd.get("/config/_global/stage").send().get();
+            stage = stageResponse.getNode().getValue();
+            if ("development".equals(stage)) {
+                initDevelopmentMode();
             } else {
                 logger.debug("cas started in production stage");
                 logger.debug("only installed dogus can get a ST");
                 EtcdResponsePromise<EtcdKeysResponse> response1 = etcd.getDir("/dogu").recursive().send();
                 EtcdKeysResponse response2 = etcd.get("/config/_global/fqdn").send().get();
                 fqdn = response2.getNode().getValue();
```

Comment:
response1 and response2 a very bad variable names, rename them to something listDoguResponse and and fqdnResponse.